src/components/ArticleCard.jsx

```diff
@@ -6,8 +6,55 @@ import Typography from "@mui/material/Typography";
 import Grid from "@mui/material/Grid";
 import dayjs from "dayjs";
 import { Link } from "react-router-dom";
+import ThumbUpIcon from "@mui/icons-material/ThumbUp";
+import { ThumbDown } from "@mui/icons-material";
+import { Button } from "@mui/material";
+import { patchArticleById } from "../api/api";
+import { useState } from "react";
 
 export default function ArticleCard({ article }) {
+  const [optimisticVotes, setOptimisticVotes] = useState(0);
+  const [isUpvoted, setIsUpvoted] = useState(false);
+  const [isDownvoted, setIsDownvoted] = useState(false);
+  const [err, setErr] = useState(null);
+
+  const handleUpvoteClick = (id, isUpvote) => {
+    if (!isUpvoted) {
+      setOptimisticVotes((currVotes) => {
+        setErr(null);
+        return currVotes + 1;
+      });
+      patchArticleById(id, isUpvote).catch((err) => {
+        setOptimisticVotes((currVotes) => {
+          setErr("Something went wrong, please try again.");
+          return currVotes - 1;
+        });
+      });
+    } else {
+      alert("You may only vote once");
+    }
+    setIsUpvoted(true);
+  };
+
+  const handleDownvoteClick = (id, isDownvote) => {
+    if (!isDownvoted) {
+      setOptimisticVotes((currVotes) => {
+        setErr(null);
+        return currVotes - 1;
+      });
+      patchArticleById(id, isDownvote).catch((err) => {
+        setOptimisticVotes((currVotes) => {
+          setErr("Something went wrong, please try again.");
+          return currVotes + 1;
+        });
+      });
+    } else {
+      alert("You may only vote once");
```

Comment:
it would be a nicer user experience if they knew they couldn't vote *before* they tried to vote (like by disabling the button changing the style)